### PR TITLE
[Fix] `DayPicker`, `DayPickerSingleDateController`: Snap to new date if date is off screen

### LINKED ISSUE #1320: DayPickerSingleDateController Jump to current month on `date` prop change
Here's my simple component
```javascript
    <DayPickerSingleDateController
      date={getDateFromUrl(this.props.router)}
      focused
      numberOfMonths={1}
      onDateChange={console.log}
      onFocusChange={noop}
    />
```

My URLs are in the format `/date/${yyyy}/${mm}/${dd}`, and the `getDateFromUrl()` function just transforms the URL into a momentObject.

**Repro**:
- Load page /date/2017/01/01, the DayPickerSingleDateController will highlight Jan 1st 2017.
- Navigate client side (i.e. no page refresh) to /date/2018/01/01

**Expected result:**
- DayPickerSingleDateController changes its visible month to Jan 2018

**Actual result:**
- DayPickerSingleDateController stays as Jan 2017.
- Jan 1st 2017 gets un-highlighted (this is good)
- If I manually scroll to Jan 2018, I see Jan 1st 2018 being hightlighted (which means the `date` prop has correctly been changed)

### PR DESCRIPTION
**Pull request by jayswain**
Fixes #1320 

Please review and let me know if theres any additional work you'd like completed, or if you'd like it done a different way, or, if everything is horrible 👹

### CODE REVIEW

**Comment by GoGoCarl** [comment]
This looks legit from an eyeball inspection to me! I'll pull it down sometime this week and give it some run.

**Comment by antoniomaia** [comment]
Any idea when this will be released?

**Comment by jayswain** [comment]
There was another approach that looked like was more accepted but it also fizzled out. see #1487

**Comment by viktorsec** [comment]
I'd also find this very useful!

**Comment by mparker11** [comment]
Any word on when a solution will be pushed?

File: test/components/DayPickerSingleDateController_spec.jsx [hunk 1]
```diff
@@ -46,6 +46,32 @@ describe('DayPickerSingleDateController', () => {
...
+        });
+      });
+
+      describe('date is visible', () => {
+        it('setState gets called with existing month', () => {
+          sinon.stub(isDayVisible, 'default').returns(true);
+          const date = today;
+          const newDate = moment().add(1, 'month');
+          const wrapper = shallow(<DayPickerSingleDateController {...props} date={date} />);
+          expect(wrapper.instance().state.currentMonth).to.equal(date);
```
Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx:68]:
```suggestion
          expect(wrapper.state()).to.have.property('currentMonth', date);
```

File: test/components/DayPickerSingleDateController_spec.jsx [hunk 2]
```diff
@@ -46,6 +46,32 @@ describe('DayPickerSingleDateController', () => {
...
+          wrapper.instance().componentWillReceiveProps({ ...props, date: newDate });
+          expect(wrapper.instance().state.currentMonth).to.equal(date);
```
Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx:70]:
```suggestion
          expect(wrapper.state()).to.have.property('currentMonth', date);
```

File: test/components/DayPickerSingleDateController_spec.jsx [hunk 3]
```diff
@@ -46,6 +46,32 @@ describe('DayPickerSingleDateController', () => {
...
+    describe('props.date changed', () => {
+      describe('date is not visible', () => {
+        it('setState gets called with new month', () => {
+          sinon.stub(isDayVisible, 'default').returns(false);
+          const date = today;
+          const newDate = moment().add(1, 'month');
+          const wrapper = shallow(<DayPickerSingleDateController {...props} date={date} />);
+          expect(wrapper.instance().state.currentMonth).to.equal(date);
+          wrapper.instance().componentWillReceiveProps({ ...props, date: newDate });
+          expect(wrapper.instance().state.currentMonth).to.equal(newDate);
```
Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx:58]:
```suggestion
          expect(wrapper.state()).to.have.property('currentMonth', newDate);
```

File: test/components/DayPickerSingleDateController_spec.jsx [hunk 4]
```diff
@@ -46,6 +46,32 @@ describe('DayPickerSingleDateController', () => {
...
     };
 
+    describe('props.date changed', () => {
+      describe('date is not visible', () => {
+        it('setState gets called with new month', () => {
+          sinon.stub(isDayVisible, 'default').returns(false);
+          const date = today;
+          const newDate = moment().add(1, 'month');
+          const wrapper = shallow(<DayPickerSingleDateController {...props} date={date} />);
+          expect(wrapper.instance().state.currentMonth).to.equal(date);
```
Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx:56]:
```suggestion
          expect(wrapper.state()).to.have.property('currentMonth', date);
```

**Review by ljharb — APPROVED** [review]
This looks great; I'll rebase this and land it if tests pass.

**Pull request merged**